src/Equivalence_Reasoner.thy

```diff
@@ -104,6 +104,234 @@ text_raw \<open>\label{implementation}\<close>
+  communicate information to their callers only via the goal state.
+
+  To iterate through the input inclusions, the \<^theory_text>\<open>generate_relax_inclusions\<close> method recursively
+  invokes itself. Since it already uses the goal premises for storing the shrinking set of input
+  inclusions, it stores the growing set of output inclusions in a fact argument \<^theory_text>\<open>accumulator\<close>,
+  which is amended at each recursive invocation. Users of the method should not specify this
+  argument, so that the method starts with \<^theory_text>\<open>accumulator\<close> being the empty fact list.
+\<close>
+
+method generate_relax_inclusions for R :: "'a \<Rightarrow> 'a \<Rightarrow> bool" uses accumulator = (
+  \<comment> \<open>Pick an input inclusion of the form \<open>_ \<le> R\<close>, called a base inclusion, and process it:\<close>
+  match premises in base [thin]: "S \<le> R" (cut) for S \<Rightarrow> \<open>
+    \<comment> \<open>If the base inclusion is extendable, add its extensions to the set of input inclusions:\<close>
+    (
+      \<comment> \<open>Add the extensions of the base inclusion to the set of input inclusions:\<close>
+      match premises in extensions [thin]: "_ \<le> S" (multi, cut) \<Rightarrow> \<open>
+        insert extensions [THEN order_trans, OF base]
+      \<close>
+    )?,
+    \<comment> \<open>Continue with the base inclusion added to the set of output inclusions:\<close>
+    generate_relax_inclusions R accumulator: accumulator base
+  \<close> |
+  \<comment> \<open>Remove all unused input inclusions and invalid premises:\<close>
+  (match premises in leftover [thin]: _ (multi, cut) \<Rightarrow> \<open>succeed\<close>)?,
+  \<comment> \<open>Add \<^term>\<open>R \<le> R\<close> to the set of output inclusions and return the result:\<close>
+  insert accumulator order_refl [of R]
+)
+
+text \<open>
+  We introduce a helper method \<^theory_text>\<open>relax\<close> that takes a set of inclusions and turns all goal premises
+  of the shape \<open>P\<^sub>1 \<Longrightarrow> \<cdots> \<Longrightarrow> P\<^sub>n \<Longrightarrow> S x y\<close> for which \<^term>\<open>S \<le> R\<close> is among the given conclusions
+  into \<open>P\<^sub>1 \<Longrightarrow> \<cdots> \<Longrightarrow> P\<^sub>n \<Longrightarrow> R x y\<close>, while dropping all other premises.
+
+  The \<^theory_text>\<open>relax\<close> method recursively invokes itself. It removes the input premises one by one from the
+  goal and collects the output premises in a fact argument \<^theory_text>\<open>accumulator\<close>. As with
+  @{method generate_relax_inclusions}, users of \<^theory_text>\<open>relax\<close> should not specify the \<^theory_text>\<open>accumulator\<close>
+  argument, so that the method starts with \<^theory_text>\<open>accumulator\<close> being the empty fact list.
+\<close>
+
+method relax uses inclusions accumulator = (
```

Comment:
The problem lies in the following `match` invocation:

https://github.com/input-output-hk/equivalence-reasoner/blob/cb48b2076761c838f4299e0b586e6ee469110f3e/src/Equivalence_Reasoner.thy#L238-L248

Apparently, method expressions are evaluated eagerly to a certain extend. Concretely, the following seems to be the case:
  * Whenever a method expression is about to be executed, all fact expressions inside it are evaluated.
  * Method expressions inside cartouches within `match` alternatives are not considered directly part of the method expressions that contain the respective `match` invocations, and thus the fact expressions inside them are evaluated only when the respective alternatives are picked.

The consequence of that regarding the above code snippet is that the expression `premise [THEN inclusion [THEN predicate2D]]` in line 247 is evaluated as soon as the enclosing `match` invocation finds an inclusion. This is before the test in lines 240–243 is run. As a result, the `relax` method tries to relax the premise before it has made sure that the current inclusion fits it. If it doesn’t, the unification error you reported arises.

A quick solution to this problem, which I will use for now, is to wrap the recursive invocation of `relax` in lines 244–247 with a dummy `match`, in order to have it executed only once this dummy `match` is run. Solving #20 will probably make this problem disappear, so that this quick fix will not be needed anymore in the longer run.

Reply:
It has turned out that the phenomenon described in #15 is also caused by eager evaluation of method expressions and thus fact expressions within them; so, unlike what I originally thought, it has nothing to do with backtracking. Since that kind of eager evaluation is probably rather a feature than a bug, I will close #15.

Reply:
Your example code actually reveals another bug. The invocation of `relax inclusions: R1_in_R2` in your *first* example turns the goal `(R1 c d ⟹ R1 a b) ⟹ R1 c d ⟹ R2 a b` into `R1 c d ⟹ (R1 c d ⟹ R2 a b) ⟹ R2 c d ⟹ R2 a b`. This is *not* intended, since it means that the unrelaxed second premise is retained.

The root of this problem seems to be that thinning doesn’t work as expected in the presence of recursion: it only seems to work for the outermost invocation of the recursive method. I will investigate that further and come up with a solution.

Reply:
The “no unifiers” issue is now resolved in 9a0ebeff82336184cd4217f6d901f943fecfa52f.

Reply:
I think I finally know why `relax` retains unrelaxed premises (in fact, all unrelaxed premises but the first one). There seems to be a bug in Eisbach, but this bug has nothing to do with recursion, as I had conjectured [above](https://github.com/input-output-hk/equivalence-reasoner/pull/6#discussion_r774242579). The apparent nature of this bug is now described in #24.

Reply:
The problem of premises being retained is now resolved in fa11cff7f5c7fb798e07082cfb522fee6f7f2823.

Solving #20 will hopefully make this kind of fix unnecessary.